sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/perPartitionAutomaticFailover/GlobalPartitionEndpointManagerForPerPartitionAutomaticFailover.java

```diff
@@ -0,0 +1,166 @@
+            return false;
+        }
+
+        checkNotNull(request, "Argument 'request' cannot be null!");
+        checkNotNull(request.requestContext, "Argument 'request.requestContext' cannot be null!");
+
+        if (request.isReadOnlyRequest()) {
+            return false;
+        }
+
+        if (!isPerPartitionAutomaticFailoverApplicable(request)) {
+            return false;
+        }
+
+        PartitionKeyRange partitionKeyRange = request.requestContext.resolvedPartitionKeyRangeForPerPartitionAutomaticFailover;
+        String resolvedCollectionRid = request.requestContext.resolvedCollectionRid;
+
+        if (partitionKeyRange == null) {
+            return false;
+        }
+
+        if (StringUtils.isEmpty(resolvedCollectionRid)) {
+            return false;
+        }
+
+        PartitionKeyRangeWrapper partitionKeyRangeWrapper = new PartitionKeyRangeWrapper(partitionKeyRange, resolvedCollectionRid);
+        RegionalRoutingContext failedRegionalRoutingContext = request.requestContext.regionalRoutingContextToRoute;
+
+        if (failedRegionalRoutingContext == null) {
+            return false;
+        }
+
+        PartitionLevelFailoverInfo partitionLevelFailoverInfo
+            = this.partitionKeyRangeToLocation.computeIfAbsent(partitionKeyRangeWrapper, partitionKeyRangeWrapper1 -> new PartitionLevelFailoverInfo(failedRegionalRoutingContext, this.globalEndpointManager));
+
+        // Rely on account-level read endpoints for new write region discovery
+        List<RegionalRoutingContext> accountLevelReadRoutingContexts = this.globalEndpointManager.getAvailableReadRoutingContexts();
+
+        if (partitionLevelFailoverInfo.tryMoveToNextLocation(accountLevelReadRoutingContexts, failedRegionalRoutingContext)) {
+
+            request.requestContext.setPerPartitionAutomaticFailoverInfoHolder(partitionLevelFailoverInfo);
+            return true;
+        }
+
+        this.partitionKeyRangeToLocation.remove(partitionKeyRangeWrapper);
+        return false;
+    }
+
+    public boolean isPerPartitionAutomaticFailoverEnabled() {
+        return this.isPerPartitionAutomaticFailoverEnabled;
+    }
+
+    public boolean isPerPartitionAutomaticFailoverApplicable(RxDocumentServiceRequest request) {
+
+        if (!this.isPerPartitionAutomaticFailoverEnabled) {
+            return false;
+        }
+
+        if (this.globalEndpointManager.getApplicableReadEndpoints(Collections.emptyList()).size() <= 1) {
+            return false;
+        }
+
+        checkNotNull(request, "Argument 'request' cannot be null!");
+
+        ResourceType resourceType = request.getResourceType();
+        OperationType operationType = request.getOperationType();
+
+        checkNotNull(resourceType, "Argument 'resourceType' cannot be null!");
+        checkNotNull(operationType, "Argument 'operationType' cannot be null!");
+
+        if (request.getOperationType() == OperationType.QueryPlan) {
```

Comment:
same here?

Reply:
We don't need the `ResourceType.Document` check here since `QueryPlan` calls aren't eligible in itself.